src/main/java/net/snowflake/client/core/bind/BindUploader.java

```diff
@@ -0,0 +1,438 @@
+  {
+    ArrayList<ArrayList<String>> columns = new ArrayList<>(bindValues.size());
+    for (int i = 1; i <= bindValues.size(); i++)
+    {
+      // bindValues should have n entries with string keys 1 ... n and list values
+      String key = Integer.toString(i);
+      if (!bindValues.containsKey(key))
+      {
+        throw new BindException(
+            String.format("Bind map with %d columns should contain key \"%d\"", bindValues.size(), i));
+      }
+
+      ParameterBindingDTO value = bindValues.get(key);
+      try
+      {
+        ArrayList<String> list = (ArrayList<String>) value.getValue();
+        columns.add(i-1, list);
+      }
+      catch (ClassCastException ex)
+      {
+        throw new BindException("Value in binding DTO could not be cast to a list");
+      }
+    }
+    return columns;
+  }
+
+  /**
+   * Transpose a list of columns and their values to a list of rows
+   * @param columns
+   * @return list of rows
+   * @throws BindException
+   */
+  private List<String[]> buildRows(ArrayList<ArrayList<String>> columns) throws BindException
+  {
+    List<String[]> rows = new ArrayList<>();
+
+    int numColumns = columns.size();
+    // columns should have binds
+    if (columns.get(0).isEmpty())
+    {
+      throw new BindException("No binds found in first column");
+    }
+
+    int numRows = columns.get(0).size();
+    // every column should have the same number of binds
+    for (int i = 0; i < numColumns; i++)
+    {
+      int iNumRows = columns.get(i).size();
+      if (columns.get(i).size() != numRows)
+      {
+        throw new BindException(
+            String.format("Column %d has a different number of binds (%d) than column 1 (%d)", i, iNumRows, numRows));
+      }
+    }
+
+    for (int rowIdx = 0; rowIdx < numRows; rowIdx++)
+    {
+      String[] row = new String[numColumns];
+      for (int colIdx = 0; colIdx < numColumns; colIdx++)
+      {
+        row[colIdx] = columns.get(colIdx).get(rowIdx);
+      }
+      rows.add(row);
+    }
+
+    return rows;
+  }
+
+  /**
+   * Write the list of rows to compressed CSV files in the temporary directory
+   * @param rows
+   * @throws BindException
+   */
+  private void writeRowsToCSV(List<String[]> rows) throws BindException
+  {
+    int numBytes = 0;
+    int rowNum = 0;
+    int fileCount = 0;
+
+    while(rowNum < rows.size())
+    {
+      File file = getFile(++fileCount);
+
+      try (OutputStream out = openFile(file))
+      {
+        // until we reach the last row or the file is too big, write to the file
+        numBytes = 0;
+        while(numBytes < fileSize && rowNum < rows.size())
+        {
+          byte[] csv = createCSVRecord(rows.get(rowNum));
+          numBytes += csv.length;
+          out.write(csv);
+          rowNum++;
+        }
+      }
+      catch (IOException ex)
+      {
+        throw new BindException(
+            String.format("Exception encountered while writing to file: %s", ex.getMessage()));
+      }
```

Comment:
Is outputStream closed? Maybe a `final` block to close the stream?

Reply:
The `try(...)` block implicitly closes the resource allocated inside the parentheses on exiting the block (even on exception).